### PR TITLE
Introduce fixed elevation range for raster layers

### PR DESCRIPTION
**Pull request by nyalldawson**
This introduces a new option for specifying how raster layers have associated elevation. It permits a fixed elevation range to be set for the layer. It can be used when the layer has a single fixed elevation, or a range (slice) of elevation values.

Users can set the lower and upper elevation range for the layer, and whether the lower or upper limits are inclusive or exclusive.

When enabled, the layer will only be visible in elevation filtered 2d maps when the layer's range is included in the map's z range. (Elevation profile support for this mode will come in a follow up PR)

image

### CODE REVIEW

File: src/core/raster/qgsrasterlayerelevationproperties.cpp [hunk 1]
```diff
@@ -150,15 +222,25 @@ void QgsRasterLayerElevationProperties::setBandNumber( int band )
...
+    case Qgis::RasterElevationMode::FixedElevationRange:
+      return mFixedRange;
 
-  return pixelValue * mZScale + mZOffset;
+    case Qgis::RasterElevationMode::RepresentsElevationSurface:
+    {
+      if ( band != mBandNumber )
+        return QgsDoubleRange();
+
+      return QgsDoubleRange( pixelValue * mZScale + mZOffset, pixelValue * mZScale + mZOffset );
```
Comment (nirvn) [inline src/core/raster/qgsrasterlayerelevationproperties.cpp:240]: Micro optimization: avoid counting it twice by using a const double?

**Review by wonder-sk — APPROVED** [review]
Looks good!

File: src/app/raster/qgsrasterelevationpropertieswidget.cpp [hunk 2]
```diff
@@ -118,9 +181,60 @@ void QgsRasterElevationPropertiesWidget::apply()
...
+      break;
+    case Qgis::RangeLimits::ExcludeLowerIncludeUpper:
+      includeLower = false;
+      break;
+    case Qgis::RangeLimits::ExcludeBoth:
+      includeLower = false;
+      includeUpper = false;
+      break;
+  }
+  props->setFixedRange( QgsDoubleRange( fixedLower, fixedUpper, includeLower, includeUpper ) );
```
[the remark this answers is not in the document]
Comment (nyalldawson) [reply src/app/raster/qgsrasterelevationpropertieswidget.cpp:209]: Good idea, done

File: src/core/raster/qgsrasterlayerelevationproperties.cpp (hunk 1 above)
Comment (nyalldawson) [reply src/core/raster/qgsrasterlayerelevationproperties.cpp]: Ok, done

**Pull request merged**